### PR TITLE
[mlir][EraVM] Implement bytecode generation

### PR DESCRIPTION
**Pull request by abinavpp**
# What ❔

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.

### CODE REVIEW

File: test/lit/mlirCodegen/EraVM/bytecode-gen.sol [hunk 1]
```diff
@@ -0,0 +1,8 @@
...
...0008000000d103550009000000d10355000a000000d10355000b000000d10355000c000000d10355000d000000d10355000e000000d10355000f000000d103550010000000d103550011000000d103550012000000d10355001d00000002001f001300000003001f001400000004001f001500000005001f001600000006001f001700000007001f001800000008001f001900000009001f001a0000000a001f001b0000000b001f001c0000000c001f000000010120018f000000000101004b000000750000613d000000730000013d009500010000040f000000770000013d009500150000040f000000770000013d0000002a01000039000000000001042d0001000000000002000100000003001d00000000030200190000000102000029000000270410009c0000002701008041000000270430009c000000270300804100000040011002100000006003300210000000e002200210000000000131019f000000000121019f00000097000104300001000000000002000100000003001d00000000030200190000000102000029000000270410009c0000002701008041000000270430009c000000270300804100000040011002100000006003300210000000e002200210000000000131019f000000000121019f000000960001042e0000009500000432000000960001042e0000009700010430000000000000000000000000000000000000000000000000000000006808fd0700000000000000000000000000000000000000000000000000000000ffffffff0000000000000000000000000000000000000000000000000000000000000000
```
Comment (hedgar2017) [inline test/lit/mlirCodegen/EraVM/bytecode-gen.sol]: Is it a strict check?
Will break after merging an optimization.

Comment (abinavpp) [reply test/lit/mlirCodegen/EraVM/bytecode-gen.sol]: I added this test just for this draft. I don't intend to keep it in the main branch.

Comment (abinavpp) [reply test/lit/mlirCodegen/EraVM/bytecode-gen.sol]: This output matches that of zksolc's `--bin` output. For now, would this work for the compiler-tester integration? Or should I wrap it in json?

File: libsolidity/codegen/mlir/Passes.cpp [hunk 2]
```diff
@@ -194,12 +199,38 @@ bool solidity::mlirgen::doJob(JobSpec const &job, mlir::MLIRContext &ctx,
...
     llvmPassMgr.run(*llvmMod);
+
+    if (job.action == Action::PrintAsm) {
+      llvm::outs() << outStream.str();
+
+      // Print the bytecode in hex.
+    } else {
+      auto obj = llvm::MemoryBuffer::getMemBuffer(
+          /*InputData=*/outStream.str(), /*BufferName=*/"",
+          /*RequiresNullTerminator=*/false);
```
Comment (PavelKopyl) [inline libsolidity/codegen/mlir/Passes.cpp]: You can also create`obj` as following:
```
 LLVMMemoryBufferRef obj = LLVMCreateMemoryBufferWithMemoryRange(
      outStream.data(), outStream.size(), "Input", /*RequiresNullTerminator=*/0);
```
Then you will not need to use `llvm::wrap` below.

File: libsolidity/codegen/mlir/Passes.cpp [hunk 3]
```diff
@@ -194,12 +199,38 @@ bool solidity::mlirgen::doJob(JobSpec const &job, mlir::MLIRContext &ctx,
...
+      LLVMMemoryBufferRef cBindingBytecode = nullptr;
+      char *errMsg = nullptr;
+      if (LLVMLinkEraVM(llvm::wrap(obj.get()), &cBindingBytecode,
+                        /*metadataPtr=*/nullptr, /*metadataSize=*/0, &errMsg)) {
+        llvm_unreachable(errMsg);
```
Comment (PavelKopyl) [inline libsolidity/codegen/mlir/Passes.cpp:224]: Please, also note that theoretically the `errMsg` should be disposed, as `LLVMLinkEraVM` allocates a memory buffer for the massage. For this you can use `LLVMDisposeMessage(errMsg)`. But as long as `llvm_unreachable` is called here, this has no meaning.

File: test/lit/mlirCodegen/EraVM/bytecode-gen.sol (hunk 1 above)
Comment (hedgar2017) [reply test/lit/mlirCodegen/EraVM/bytecode-gen.sol]: compiler-tester needs full standard JSON output like that of vanilla solc.
Replacing the EVM bytecode would be enough for now.

File: libsolidity/codegen/mlir/Passes.cpp (hunk 2 above)
Comment (abinavpp) [reply libsolidity/codegen/mlir/Passes.cpp]: Thanks!

**Pull request closed without merging**